Read only endpoint for XCom #8134

randr97 opened the pull request:
Closes #8134 
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

mik-laj commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -872,7 +872,9 @@ paths:
+  /dags/{dag_id}/taskInstances/{task_id}/{execution_date}/xcomEntries/{xcom_key}:
     parameters:
       - $ref: '#/components/parameters/DAGID'
       - $ref: '#/components/parameters/DAGRunID'
```
Can you update the spec here also?

mik-laj commented on airflow/api_connexion/openapi/v1.yaml:
```diff
@@ -872,7 +872,9 @@ paths:
           $ref: '#/components/responses/PermissionDenied'
 
-  /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/xcomEntries/{xcom_key}:
+  # TODO: /dags/{dag_id}/dagRuns/{dag_run_id}/taskInstances/{task_id}/xcomEntries/{xcom_key}:
```
Can you add a description that will contain the necessary steps to take and the reason for the change?

randr97 committed:
Read only endpoint for XCom #8134

ashb committed:
Add SQLA Relationship between DagRun and TI

This lets us select like this:

```
(
    s.query(DagRun)
    .filter_by(dag_id=='example_dag1', run_id=='scheduled__2020-06-08T11:00:00+00:00')
    .join(DagRun.task_instances)
    .with_entities(TaskInstance).filter_by(task_id='print_date2')
```

This lets us easily go from the URL arguments of the API to the task
instance(s)

mik-laj commented on airflow/api_connexion/schemas/xcom_schema.py:
```diff
@@ -0,0 +1,67 @@
+        model = XCom
+
+    COLLECTION_NAME = 'xcom_entries'
+    FIELDS_FROM_NONE_TO_EMPTY_STRING = ['key', 'task_id', 'dag_id']
```
These fields are primary keys, so they must always have a value. Why do you want them to be empty?

randr97 replied:
yep realised, will remove them in my next draft

Commit:
Use explicit join between xcom and dag_run

ephraimbuddy commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -26,18 +31,35 @@ def delete_xcom_entry():
+
+    q_object = query.one_or_none()
+    if not q_object:
+        raise Exception("Object Not found")
```
```suggestion
        raise NotFound("Xcom not found")
```

OmairK replied:
API is exceptions are defined in one of the PRs [here](https://github.com/apache/airflow/blob/287932c276dae6bce242051a40cff3680c238626/airflow/api_connexion/exceptions.py#L20). I think it would be better to raise a 404 exception here. @randr97  what do you think ?

ephraimbuddy commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -15,8 +15,13 @@
-# TODO(mik-laj): We have to implement it.
-#     Do you want to help? Please look at: sshttps://github.com/apache/airflow/issues/8134
+from sqlalchemy import and_
+
```
```suggestion
from airflow.api_connexion.exceptions import NotFound
```

ephraimbuddy commented on tests/api_connexion/endpoints/test_xcom_endpoint.py:
```diff
@@ -18,59 +18,153 @@
+        with create_session() as session:
+            session.query(XCom).delete()
+            session.query(DR).delete()
 
```
I suggest you add tearDown method so that there's no hanging Xcom and DR data after the tests

ephraimbuddy commented on tests/api_connexion/endpoints/test_xcom_endpoint.py:
```diff
@@ -18,59 +18,153 @@
+    def test_should_response_200(self, session):
+        dag_id = 'test-dag-id'
+        task_id = 'test-task-id'
+        execution_date = '2005-04-02T21:37:42+00:00'
```
```suggestion
        execution_date = '2005-04-02T21:00:00+00:00'
```
The date will be readable this way

mik-laj replied:
@randr97 Can you fix it?

ephraimbuddy commented on tests/api_connexion/schemas/test_xcom_schema.py:
```diff
@@ -0,0 +1,193 @@
+class TestXComCollectionItemSchema(unittest.TestCase):
+
+    def setUp(self) -> None:
+        self.now = timezone.utcnow()
```
Can you use a date string here. It will help in review. Also you may want to refactor `now` to `default_time`

ephraimbuddy commented on tests/api_connexion/schemas/test_xcom_schema.py:
```diff
@@ -0,0 +1,193 @@
+    def test_serialize(self, session):
+        xcom_model = XCom(
+            key='test_key',
+            timestamp=self.now,
```
If you use a date string above, you can parse the time with timezone.parse(timestring)

ephraimbuddy commented on tests/api_connexion/schemas/test_xcom_schema.py:
```diff
@@ -0,0 +1,193 @@
+            deserialized_xcom[0],
+            {
+                'key': 'test_key',
+                'timestamp': self.now.isoformat(),
```
There will be no need for isoformat() when you refactor

ephraimbuddy commented on tests/api_connexion/schemas/test_xcom_schema.py:
```diff
@@ -0,0 +1,193 @@
+        self.now = timezone.utcnow()
+        with create_session() as session:
+            session.query(XCom).delete()
+
```
Would be nice to add a tearDown here

houqp commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -26,18 +32,35 @@ def delete_xcom_entry():
 
-def get_xcom_entries():
+@provide_session
+def get_xcom_entries(dag_id, dag_run_id, task_id, session):
```
can you add type annotation to all handler functions?

ephraimbuddy replied:
Hi @houqp , Is it necessary to add type annotation since connexion will be using it internally?

houqp replied:
@ephraimbuddy yes, type annotation helps catching bugs at build time for internal and external code.

mik-laj replied:
+1

houqp commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -26,18 +32,35 @@ def delete_xcom_entry():
+                              XCom.task_id == task_id))
+    query = query.join(DR, and_(XCom.dag_id == DR.dag_id, XCom.execution_date == DR.execution_date))
+    query = query.filter(DR.run_id == dag_run_id)
+    return xcom_collection_schema.dump(XComCollection(xcom_entries=query.all(), total_entries=query.count()))
```
does `query.count()` issue another count query to the db? if so, we should call `len` on rows returned by `query.all()` instead.

mik-laj replied:
The total_entries field should contain the number of all items on all pages. This allows us to build pagination and build the user interface more easily.

randr97 replied:
Then there should be some order-by if I'm not wrong @mik-laj . Can we look at this - https://github.com/wizeline/sqlalchemy-pagination 
What is ur take on this?

mik-laj commented on airflow/models/dagrun.py:
```diff
@@ -66,6 +66,13 @@ class DagRun(Base, LoggingMixin):
         UniqueConstraint('dag_id', 'run_id'),
     )
 
+    task_instances = relationship(
```
Is it still used?

randr97 replied:
It's not used...prolly will remove it.
@mik-laj  what is ur opinion, do we have it with a TODO sorta comment??

mik-laj commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -26,18 +33,41 @@ def delete_xcom_entry():
+    query = query.filter(DR.run_id == dag_run_id).order_by(
+        XCom.execution_date, XCom.task_id, XCom.dag_id, XCom.key
+    )
+    total_entries = query.count()
```
Can you use optimized query here?
https://github.com/apache/airflow/pull/9361

mik-laj commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -26,18 +33,41 @@ def delete_xcom_entry():
+    offset = request.args.get(parameters.page_offset, 0)
+    limit = min(int(request.args.get(parameters.page_limit, 100)), 100)
+    query = session.query(XCom)
+    query = query.filter(and_(XCom.dag_id == dag_id,
```
```
This endpoint allows specifying `~` as the dag_id, dag_run_id, task_id to retrieve XCOM entries for
        for all DAGs, DAG Runs and task instances.
```
Can you add support for it?

mik-laj commented on tests/api_connexion/endpoints/test_xcom_endpoint.py:
```diff
@@ -15,62 +15,254 @@
+    def test_should_response_200(self, session):
+        dag_id = 'test-dag-id'
+        task_id = 'test-task-id'
+        execution_date = '2005-04-02T21:37:42+00:00'
```
Here too.

mik-laj commented on tests/api_connexion/endpoints/test_xcom_endpoint.py:
```diff
@@ -15,62 +15,254 @@
+        parsed = dict(parse.parse_qsl(parse.urlsplit(url).query))
+        limit = int(parsed.get('limit', 100))
+        offset = int(parsed.get('offset', 0))
+        expected_xcom_ids = self.expected_xcom_ids[0 + offset: offset + limit]
```
This should be a constant value. Whenever possible, the tests should only create variables, call other functions, do assertions. They should not calculate anything because you could have made a mistake in two places. 

https://testing.googleblog.com/2019/12/testing-on-toilet-tests-too-dry-make.html

mik-laj commented on airflow/api_connexion/endpoints/xcom_endpoint.py:
```diff
@@ -26,18 +35,58 @@ def delete_xcom_entry():
+
+    query_object = query.one_or_none()
+    if not query_object:
+        raise NotFound("XCom object not found")
```
```suggestion
        raise NotFound("XCom entry not found")
```

mik-laj commented on tests/api_connexion/schemas/test_xcom_schema.py:
```diff
@@ -0,0 +1,207 @@
+        self.default_time = '2005-04-02T21:00:00+00:00'
+        self.default_time_parsed = parse_execution_date(self.default_time)
+        with create_session() as session:
+            session.query(XCom).delete()
```
```suggestion
```

mik-laj commented on tests/api_connexion/schemas/test_xcom_schema.py:
```diff
@@ -0,0 +1,207 @@
+class TestXComCollectionItemSchema(TestXComSchemaBase):
+
+    def setUp(self) -> None:
+        self.default_time = '2005-04-02T21:00:00+00:00'
```
```suggestion
        suoer().setUp()
        self.default_time = '2005-04-02T21:00:00+00:00'
```

mik-laj commented on airflow/api_connexion/schemas/xcom_schema.py:
```diff
@@ -0,0 +1,63 @@
+
+xcom_schema = XComSchema()
+xcom_collection_item_schema = XComCollectionItemSchema()
+xcom_collection_schema = XComCollectionSchema()
```
```suggestion
xcom_schema = XComSchema(strict=True)
xcom_collection_item_schema = XComCollectionItemSchema(strict=True)
xcom_collection_schema = XComCollectionSchema(strict=True)
```
This is the last thing. If you accept this, I will merge this change.

randr97 committed:
Read only endpoint for XCom #8134
